### DISCUSSION TITLE
how to increase tflops utilization

### DISCUSSION THREAD

**Discussion by pathfinder-pf** [jax-ml/jax#35601] [Q&A]
I have two matrices A and B, with shapes (m,k) and (k,n) respectively. I'm performing matmul calculations on A and B. I found that the tflops is relatively low when k=128, but increases as k increases. How can I improve the tflops with a fixed k, for example, k=128? Below are my tests on a TPU v6e
````
import jax
import time
import jax.numpy as jnp
import numpy as np
t=jnp.bfloat16
max_tflops=918

seq_q = 17280
head_dims = [128, 256, 512, 1024, 2048]
seq_kv = 17280

key = jax.random.PRNGKey(1)
key1, key2 = jax.random.split(key, num=2)

@jax.jit
def compute(x, y, z):
    tmp=jax.lax.dot(x, y)
    result=jax.lax.dot(tmp, z)
    return result

for head_dim in head_dims:
    p = jax.random.normal(key ,(seq_q, head_dim), dtype=t)
    q = jax.random.normal(key1, (head_dim, seq_kv), dtype=t)
    r = jax.random.normal(key2, (seq_kv, head_dim), dtype=t)    
    # warmup
    for i in range(3):
        y = compute(p, q, r)
        jax.block_until_ready(y)
    start_time = time.perf_counter()
    y=compute(p, q, r)
    jax.block_until_ready(y)
    end_time = time.perf_counter()
    tflops = 4 * seq_q * head_dim * seq_kv / (time.perf_counter() - start_time) / 1e12
    utilization = tflops / max_tflops
    print(f"{seq_q}, {head_dim}, {seq_kv}  {tflops} tflops, utilization {utilization}")

# 17280, 128, 17280  297.9989102154806 tflops, utilization 0.32461754925433617
# 17280, 256, 17280  579.4730728243075 tflops, utilization 0.6312342841223393
# 17280, 512, 17280  673.4684130700865 tflops, utilization 0.7336257222985691
# 17280, 1024, 17280  733.6798877979321 tflops, utilization 0.799215564050035
# 17280, 2048, 17280  734.6405938854358 tflops, utilization 0.8002620848425227
